aleph-client/src/contract/mod.rs

```diff
@@ -0,0 +1,144 @@
+use substrate_api_client::{compose_extrinsic, GenericAddress, XtStatus};
+
+use crate::{try_send_xt, AnyConnection, Connection, SignedConnection};
+
+/// Represents a contract instantiated on the chain.
+pub struct ContractInstance {
+    address: AccountId32,
+    ink_project: InkProject,
+}
+
+impl ContractInstance {
+    const MAX_READ_GAS: u64 = 500000000000u64;
+    const MAX_GAS: u64 = 10000000000u64;
+
+    /// Creates a new contract instance under `address` with metadata read from `metadata_path`.
+    pub fn new(address: AccountId32, metadata_path: &str) -> Result<Self> {
+        Ok(Self {
+            address,
+            ink_project: load_metadata(metadata_path)?,
+        })
+    }
+
+    /// The address of this contract instance.
+    pub fn address(&self) -> &AccountId32 {
+        &self.address
+    }
+
+    /// The metadata of this contract instance.
+    pub fn ink_project(&self) -> &InkProject {
+        &self.ink_project
+    }
+
+    /// Reads the value of a read-only, 0-argument call via RPC.
+    pub fn contract_read0(&self, conn: &Connection, message: &str) -> Result<Value> {
+        self.contract_read(conn, message, &[])
+    }
+
+    /// Reads the value of a read-only call via RPC.
+    pub fn contract_read(&self, conn: &Connection, message: &str, args: &[&str]) -> Result<Value> {
+        let payload = self.encode(message, args)?;
+        let request = self.contract_read_request(&payload);
+        let response = conn
+            .get_request(request)?
+            .context("RPC request error - there may be more info in node logs.")?;
+        let response_data = from_str::<serde_json::Value>(&response)?;
+        let hex_data = response_data["result"]["Ok"]["data"]
+            .as_str()
+            .context("Contract response data not found - the contract address might be invalid.")?;
+        self.decode_response(message, hex_data)
+    }
+
+    /// Executes a contract call.
+    pub fn contract_exec(
+        &self,
+        conn: SignedConnection,
+        message: &str,
+        args: &[&str],
+    ) -> Result<()> {
+        let data = self.encode(message, args)?;
+        let xt = compose_extrinsic!(
+            conn.as_connection(),
+            "Contracts",
+            "call",
+            GenericAddress::Id(self.address.clone()),
+            Compact(0u64),
+            Compact(Self::MAX_GAS),
+            None::<Option<u128>>,
+            data
+        );
+
+        try_send_xt(&conn, xt, Some("Contracts call"), XtStatus::InBlock)
+            .context("Failed to exec contract message")?;
+        Ok(())
+    }
+
+    fn contract_read_request(&self, payload: &[u8]) -> serde_json::Value {
+        let payload = hex::encode(payload);
+        json!({
+            "jsonrpc": "2.0",
+            "method": "contracts_call",
+            "params": [{
+                "origin": self.address,
+                "dest": self.address,
+                "value": 0,
+                "gasLimit": Self::MAX_READ_GAS,
```

Comment:
is this true: since it is an RPC call, we could pass here infinity?

Reply:
Dunno about infinity, but numbers larger than what's currently set "don't work". I took the value from polkadot.js and never looked back.